### PR TITLE
Add EMF concurrency test to check for log corruption.

### PR DESCRIPTION
**Pull request by jefchien**
# Description of the issue
There was an issue with the TCP log receiver corrupting logs on high concurrency. This was fixed as part of  https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/72, but was not pulled into the agent until https://github.com/aws/amazon-cloudwatch-agent/pull/857. This was undetected because it was verified using manual testing.

# Description of changes
Add an EMF concurrency test that creates multiple TCP connections to the agent and writes to them concurrently. Runs a log query to determine if any of the logs were corrupted.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Failure from v1.300028.1 (https://github.com/aws/amazon-cloudwatch-agent/commit/6b2589156bc9f27a2c7cb5174073b466163dbd7e) run
https://github.com/aws/amazon-cloudwatch-agent/actions/runs/6304504010/job/17116940882

Success from latest (https://github.com/aws/amazon-cloudwatch-agent/commit/2a3fdf9d57bf8039990156031e205a4d7e495b13) run
https://github.com/aws/amazon-cloudwatch-agent/actions/runs/6305043334/job/17117674411

### CODE REVIEW

File: test/emf_concurrent/emf_concurrent_test.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+	for _, stream := range awsservice.GetLogStreams(e.logGroupName) {
+		gotStreamNames = append(gotStreamNames, *stream.LogStreamName)
+	}
+	assert.Lenf(t, gotStreamNames, 1, "Detected corruption: multiple streams found")
+	qs := queryString()
+	log.Printf("Starting query for log group (%s): %s", e.logGroupName, qs)
+	gotLogQueryStats, err := awsservice.GetLogQueryStats(e.logGroupName, startTime.Unix(), endTime.Unix(), qs)
+	require.NoError(t, err, "Unable to get log query stats")
+	assert.NotZero(t, gotLogQueryStats.RecordsScanned, "No records found in CloudWatch Logs")
+	assert.Zerof(t, gotLogQueryStats.RecordsMatched, "Detected corruption: %v/%v records matched", gotLogQueryStats.RecordsMatched, gotLogQueryStats.RecordsScanned)
```
Comment (movence) [inline test/emf_concurrent/emf_concurrent_test.go:88]: I don't have full context of that corruption issue, but is 0 match expected or < what's been scanned?

File: test/emf_concurrent/emf_concurrent_test.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+			log.Fatalf("unable to connect to address (%s): %v", emfAddress, err)
+		}
+		conns = append(conns, conn)
+	}
+
+	log.Printf("Starting EMF emitters for log group (%s)/stream (%s)", e.logGroupName, e.logStreamName)
+	startTime := time.Now()
+	for i := 0; i < threadCount; i++ {
+		e.wg.Add(1)
+		go e.start(conns[i%len(conns)])
```
Comment (movence) [inline test/emf_concurrent/emf_concurrent_test.go:69]: nit:probably increase the # of threads to take advantage of this?

File: generator/test_case_generator.go [hunk 3]
```diff
@@ -32,6 +32,7 @@ type matrixRow struct {
 	UseSSM              bool   `json:"useSSM"`
 	ExcludedTests       string `json:"excludedTests"`
 	MetadataEnabled     string `json:"metadataEnabled"`
+	MaxAttempts         int    `json:"max_attempts"`
```
Comment (sethAmazon) [inline generator/test_case_generator.go:35]: Why is this needed if it is 1?

File: test/emf_concurrent/emf_concurrent_test.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+	defer awsservice.DeleteLogGroup(e.logGroupName)
+
+	tcpAddr, err := net.ResolveTCPAddr("tcp", emfAddress)
+	if err != nil {
+		log.Fatalf("invalid tcp emfAddress (%s): %v", emfAddress, err)
+	}
+
+	var conns []*net.TCPConn
+	for i := 0; i < connectionCount; i++ {
```
Comment (sethAmazon) [inline test/emf_concurrent/emf_concurrent_test.go:56]: Can this be done with the sdk? No sure since the sdk will open a single conn so all threads push to a single conn. Maybe we can open up multiple instances of the sdk. Also, ik there is a java emf sdk but not sure about golang.

File: generator/test_case_generator.go (hunk 3 above)
Comment (jefchien) [reply generator/test_case_generator.go:35]: Because the default is 3.

File: test/emf_concurrent/emf_concurrent_test.go (hunk 1 above)
Comment (jefchien) [reply test/emf_concurrent/emf_concurrent_test.go:88]: If any records match the query, then there's a problem. It should always be 0.

File: test/emf_concurrent/emf_concurrent_test.go (hunk 4 above)
Comment (jefchien) [reply test/emf_concurrent/emf_concurrent_test.go:56]: There's no official Go SDK for EMF or none that I could find.

File: test/emf_concurrent/emf_concurrent_test.go (hunk 2 above)
Comment (jefchien) [reply test/emf_concurrent/emf_concurrent_test.go:69]: Changed it to 15 threads/5 connections.

Before fix: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/6315736924/job/17148887131
After fix: https://github.com/aws/amazon-cloudwatch-agent/actions/runs/6316331462/job/17150704010

**Review by sethAmazon — APPROVED** [review]

File: test/emf_concurrent/emf_concurrent_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	for i := 0; i < threadCount; i++ {
+		e.wg.Add(1)
+		go e.start(conns[i%len(conns)])
+	}
+	time.Sleep(testRuntime)
+	close(e.done)
+	e.wg.Wait()
+	common.StopAgent()
+	endTime := time.Now()
+	log.Println("Stopping EMF emitters")
```
Comment (adam-mateen) [inline test/emf_concurrent/emf_concurrent_test.go:76]: nit: shouldn't this log line be before closing `e.done`?
```
log.Println("Stopping EMF emitters")
```
Probably doesn't matter since closing and waiting shouldn't take long.

File: test/emf_concurrent/emf_concurrent_test.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	log.Printf("Starting query for log group (%s): %s", e.logGroupName, qs)
+	got, err := awsservice.GetLogQueryStats(e.logGroupName, startTime.Unix(), endTime.Unix(), qs)
+	require.NoError(t, err, "Unable to get log query stats")
+	assert.NotZero(t, got.RecordsScanned, "No records found in CloudWatch Logs")
+	assert.Zerof(t, got.RecordsMatched, "Detected corruption: %v/%v records matched", got.RecordsMatched, got.RecordsScanned)
+}
+
+// queryString creates a log query string that will match corrupted logs.
+func queryString() string {
+	return fmt.Sprintf("filter ispresent(%[1]s) and ispresent(%[2]s) and (%[1]s != %[2]s or (_aws.CloudWatchMetrics.0.Metrics.0.Unit!=%[3]q) or (_aws.CloudWatchMetrics.0.Metrics.1.Unit!=%[3]q))", metricName1, metricName2, metricUnit)
```
Comment (adam-mateen) [inline test/emf_concurrent/emf_concurrent_test.go:89]: nit: could pass in the variables to functions instead of referencing globals.

File: test/emf_concurrent/emitter.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+				Metrics: []Metric{
+					{Name: metricName1, Unit: metricUnit},
+					{Name: metricName2, Unit: metricUnit},
+				},
+			},
+		},
+	}
+}
+
+func (e *emitter) createEmfLog(metadata *Metadata) []byte {
```
Comment (adam-mateen) [inline test/emf_concurrent/emitter.go:89]: Doesn't need to done now, but I can see this code to send EMF metrics being useful outside this test case.
Might want to move it to a UTIL at some point. I saw you mentioned there was no EMF package in go to do this... yikes.

File: util/awsservice/cloudwatchlogs.go [hunk 8]
```diff
@@ -144,6 +147,40 @@ func IsLogGroupExists(logGroupName string) bool {
...
+			QueryId: output.QueryId,
+		})
+		if err != nil {
+			return nil, fmt.Errorf("failed to get query results for log group (%s): %w", logGroupName, err)
+		}
+		switch results.Status {
+		case types.QueryStatusFailed, types.QueryStatusCancelled, types.QueryStatusTimeout:
+			return nil, fmt.Errorf("unexpected query status: %v", results.Status)
+		case types.QueryStatusScheduled, types.QueryStatusRunning, types.QueryStatusUnknown:
+			time.Sleep(retryInterval)
```
Comment (adam-mateen) [inline util/awsservice/cloudwatchlogs.go:177]: Is it possible for this to infinitely retry?
Do we want to enforce a max attempts here?

File: test/emf_concurrent/emf_concurrent_test.go (hunk 6 above)
Comment (jefchien) [reply test/emf_concurrent/emf_concurrent_test.go:89]: It's very specific to this test and could probably just be a var instead of a function.

File: util/awsservice/cloudwatchlogs.go (hunk 8 above)
Comment (jefchien) [reply util/awsservice/cloudwatchlogs.go:177]: It shouldn't be possible, but I can add a max attempts.

**Review by movence — DISMISSED** [review]

**Review by movence — APPROVED** [review]

**Review by sethAmazon — APPROVED** [review]
